joss/paper.md

```diff
@@ -0,0 +1,105 @@
+---
+title: 'Downstream: efficient, constant-space implementations of stream curation algorithms.'
+tags:
+  - Python
+  - Rust
+  - C++
+authors:
+  - name: Connor Yang
+    orcid: 0009-0004-1240-2362
+    affiliation: 1
+  - name: Matthew Andres Moreno
+    orcid: 0000-0003-4726-4479
+    affiliation: 1
+affiliations:
+ - name: University of Michigan
+   index: 1
+date: 10 February 2025
+bibliography: paper.bib
+---
+
+# Summary
+
+Downstream offers efficient algorithms for the curation of continuous data streams, implemented across multiple programming languages including C++, Rust, Python, Zig, and the Cerebras Software Language. In real-world applications, data streams often exceed available memory capacity and require real-time operations on a continuous, indefinite basis [@Cordeiro2016]. While traditional approaches like circular ring buffers address this limitation by retaining only the most recent data points, Downstream maintains representative, approximate records of stream history through three novel algorithms: (1) "steady," which creates evenly spaced snapshots across the entire history; (2) "stretched," which preserves important older data points; and (3) "tilted," which prioritizes recent information. The library features extensive cross-implementation testing, automated documentation and deployment, and is available through standard package managers.
+
+# Statement of Need
+
+Efficient data stream processing is a foundational challenge in modern computing systems, where the ability to analyze continuous, high-volume data has become more prevalent [@Cordeiro2016]. Applications include sensor networks [@Eiman2003], distributed big-data processing, real-time network traffic analysis, systems log management, fraud monitoring, trading in financial markets, environmental monitoring, and astronomy, which all generate data at rates that exceed practical storage capacity while requiring analysis across varying time horizons.
+
+Downstream addresses limitations of traditional approaches to data streaming by providing algorithms that compress stream history into fixed memory capacities while preserving significant information across various temporal distributions. By exclusively operating with primitive operations and eliminating memory overhead, the framework enables efficient operations in severely resource-constrained environments including embedded systems, where available memory is often measured in kilobytes. For instance, the Cerebras WSE-2 demonstrates this constraint clearly—despite having 40 gigabytes of total on-chip memory, each individual core has access to under 50 kilobytes of memory [@Lie2023].
+
+This capability is particularly valuable for applications such as hereditary stratigraphy in large-scale phylogenetic tracking, where memory-efficient evolutionary history tracking must be maintained across distributed computing environments.
```

Comment:
The capability for...
Probably want a second sentence in this paragraph saying that downstream is being used for hstrat.